magenta-lib/src/main/scala/magenta/reporting.scala

```diff
@@ -42,134 +42,118 @@ trait MessageState {
-  lazy val startContext = null
-  lazy val finished = None
-  lazy val state = RunState.NotRunning
-}
-
-case class StartMessageState(startContext: StartContext, time: DateTime, messageId: Option[UUID] = None) extends MessageState {
+case class StartMessageState(startContext: StartContext, time: DateTime, messageId: UUID) extends MessageState {
   lazy val message = startContext.originalMessage
   lazy val finished = None
   lazy val state = RunState.Running
 }
 
-case class FinishMessageState(startContext: StartContext, finish: FinishContext, time: DateTime, messageId: Option[UUID] = None) extends MessageState {
+case class FinishMessageState(startContext: StartContext, finish: FinishContext, time: DateTime, messageId: UUID) extends MessageState {
   lazy val message = startContext.originalMessage
   lazy val finished = Some(finish)
   lazy val state = RunState.Completed
 }
 
-case class FailMessageState(startContext: StartContext, fail: FailContext, time: DateTime, messageId: Option[UUID] = None) extends MessageState {
+case class FailMessageState(startContext: StartContext, fail: FailContext, time: DateTime, messageId: UUID) extends MessageState {
   lazy val message = startContext.originalMessage
   lazy val finished = Some(fail)
   lazy val state = RunState.Failed
 }
 
-object DeployReport {
-  def wrapperToTree(node: MessageWrapper, all: List[MessageWrapper]): ReportTree = {
-    val allChildren = all.filter(_.context.parentId.exists(_ == node.messageId))
+trait DeployReport {
+  def message: Message
+  def timeString: Option[String]
+  def state: RunState.Value
+  def allMessages: Seq[MessageState]
+  def children: List[DeployReportTree]
+  def isRunning: Boolean
 
-    val isEndContextMessage = (wrapper:MessageWrapper) => wrapper.stack.top.isInstanceOf[FinishContext] ||
-                                                          wrapper.stack.top.isInstanceOf[FailContext]
+  def hasChildren: Boolean = children.nonEmpty
+  def size: Int = allMessages.size
 
-    val endOption = allChildren.filter(isEndContextMessage).map(_.stack.top.asInstanceOf[ContextMessage]).headOption
-    val children = allChildren.filterNot(isEndContextMessage)
-
-    val messageState = endOption.map { end =>
-      MessageState(node.stack.top.asInstanceOf[StartContext], end, node.stack.time)
-    }.getOrElse(MessageState(node.stack.top, node.stack.time))
-
-    ReportTree(messageState, children.map(wrapperToTree(_,all)))
-  }
-
-  def apply(list: List[MessageWrapper], title: String = "", titleTime: Option[DateTime] = None): ReportTree = {
-    val time = titleTime.getOrElse( list.headOption.map(_.stack.time).getOrElse( new DateTime() ))
-    ReportTree(Report(title, time), list.headOption.map(root => List(wrapperToTree(root,list))).getOrElse(Nil))
+  def failureMessage: Option[Fail] = {
+    allMessages.filter(_.message.getClass == classOf[Fail]).map(_.message).headOption.asInstanceOf[Option[Fail]]
```

Comment:
Not related to this PR but I wonder how could one avoid relying on `asInstanceOf` in this case?